### PR TITLE
feat: improved msg for network timeouts

### LINKED ISSUE #1922: Display HTTP timeout setting on timeout
We can hint to change the `HTTP_TIMEOUT` when a timeout is encountered and display the current value

Related:

- #1912 
- #1921

### PR DESCRIPTION
**Pull request by samypr100**
## Summary

Closes #1922

When a timeout occurs, it hints to the user to configure the `UV_HTTP_TIMEOUT` env var.

Before
```
error: Failed to download distributions
  Caused by: Failed to fetch wheel: torch==2.2.0 
  Caused by: Failed to extract source distribution
  Caused by: request or response body error: operation timed out
  Caused by: operation timed out
```

After
```
error: Failed to download distributions
  Caused by: Failed to fetch wheel: torch==2.2.0 
  Caused by: Failed to extract source distribution
  Caused by: Failed to download distribution due to network timeout. Try increasing UV_HTTP_TIMEOUT.
```

## Test Plan

Wasn't sure if we'd want a test. If we do, is there a existing mechanism or preferred approach to force a timeout to occur in tests? Maybe set the timeout to 1 and add torch as an install check (although it's possible that could become flaky)?

### CODE REVIEW

**Review by charliermarsh — APPROVED** [review]

File: crates/uv-distribution/src/distribution_database.rs [hunk 1]
```diff
@@ -45,6 +45,14 @@ pub struct DistributionDatabase<'a, Context: BuildContext + Send + Sync> {
     builder: SourceDistCachedBuilder<'a, Context>,
 }
 
+fn handle_response_errors(err: reqwest::Error) -> io::Error {
+    if err.is_timeout() {
+        io::Error::new(io::ErrorKind::TimedOut, "Failed to download distribution due to network timeout. Try increasing UV_HTTP_TIMEOUT.")
```
Comment (charliermarsh) [inline crates/uv-distribution/src/distribution_database.rs:50]: It looks like Zanie suggested `HTTP_TIMEOUT`. Do you mind changing to `HTTP_TIMEOUT`, and displaying the current value too? This is great!

Comment (samypr100) [reply crates/uv-distribution/src/distribution_database.rs:50]: Good point, I decided to use `UV_HTTP_TIMEOUT` for consitency with `registry_client.rs` first check (which seems to prefer `UV_HTTP_TIMEOUT` and `HTTP_TIMEOUT` last). I could also update that `warn_user_once` as part of this PR if we want to be fully consistent.

```rust
            let timeout = env::var("UV_HTTP_TIMEOUT").or_else(|_| env::var("UV_REQUEST_TIMEOUT")).or_else(|_| env::var("HTTP_TIMEOUT")).and_then(|value| {
                value.parse::<u64>()
                    .or_else(|_| {
                        // On parse error, warn and  use the default timeout
                        warn_user_once!("Ignoring invalid value from environment for UV_REQUEST_TIMEOUT. Expected integer number of seconds, got \"{value}\".");
                        Ok(default_timeout)
                    })
            }).unwrap_or(default_timeout);
```

Comment (zanieb) [reply crates/uv-distribution/src/distribution_database.rs:50]: I think `UV_HTTP_TIMEOUT` is good, it seems good not to recommend changing a value that other clients will respect unless you need to.

Comment (zanieb) [reply crates/uv-distribution/src/distribution_database.rs:50]: Can we say how long the timeout is currently though?

**Comment by zanieb** [comment]
Thanks for contributing!

I'd really like to explore addressing this at its root too, it seems like our timeout is not correctly configured.

Regarding testing, maybe we'd spin up a temporary webserver that just doesn't respond? That feels beyond the scope of this change though. A manual test should suffice for now.

File: crates/uv-distribution/src/distribution_database.rs (hunk 1 above)
Comment (samypr100) [reply crates/uv-distribution/src/distribution_database.rs]: Added in 65a5c3542357aa1eb69950cc0b4baa316ae82ab2 and 783a4196d09f3ba10a0294529a85528c621ee8fd. I didn't see reqwest allowing us to look at config.timeout, so I ended up exposing our timeout to the RegistryClient struct instead.

**Comment by charliermarsh** [comment]
Thanks!

**Pull request merged**